[DDP] Log unused param names under DETAIL debug mode.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62209

When `TORCH_DISTRIBUTED_DEBUG=DETAIL` is set, log names and indices of unused parameters when searching for them.

Motivation is that we have seen a couple of issues occasionally when there are errors related to parameter possibly being marked as unused when it shouldn't, this can help narrow down the root cause by explicitly logging param names that are marked as unused.

Differential Revision: [D29916085](https://our.internmc.facebook.com/intern/diff/D29916085/)

rohan-varma committed:
[DDP] Log unused param names under DETAIL debug mode.

When `TORCH_DISTRIBUTED_DEBUG=DETAIL` is set, log names and indices of unused parameters when searching for them.

Motivation is that we have seen a couple of issues occasionally when there are errors related to parameter possibly being marked as unused when it shouldn't, this can help narrow down the root cause by explicitly logging param names that are marked as unused.

Differential Revision: [D29916085](https://our.internmc.facebook.com/intern/diff/D29916085/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "c10d/Utils.hpp"
 #include <c10d/default_comm_hooks.hpp>
 #include <c10d/reducer.hpp>
```
nit: this is prior to this PR. IIUC, the order of includes should be the following?

```c++
#include <c10d/reducer.hpp>

#include "c10d/Utils.hpp"
#include <c10d/default_comm_hooks.hpp>
```

rohan-varma replied:
Thanks, fixed!

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[DDP] Log unused param names under DETAIL debug mode."


When `TORCH_DISTRIBUTED_DEBUG=DETAIL` is set, log names and indices of unused parameters when searching for them.

Motivation is that we have seen a couple of issues occasionally when there are errors related to parameter possibly being marked as unused when it shouldn't, this can help narrow down the root cause by explicitly logging param names that are marked as unused.

Differential Revision: [D29916085](https://our.internmc.facebook.com/intern/diff/D29916085/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log unused param names under DETAIL debug mode."


When `TORCH_DISTRIBUTED_DEBUG=DETAIL` is set, log names and indices of unused parameters when searching for them.

Motivation is that we have seen a couple of issues occasionally when there are errors related to parameter possibly being marked as unused when it shouldn't, this can help narrow down the root cause by explicitly logging param names that are marked as unused.

Differential Revision: [D29916085](https://our.internmc.facebook.com/intern/diff/D29916085/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log unused param names under DETAIL debug mode."


When `TORCH_DISTRIBUTED_DEBUG=DETAIL` is set, log names and indices of unused parameters when searching for them.

Motivation is that we have seen a couple of issues occasionally when there are errors related to parameter possibly being marked as unused when it shouldn't, this can help narrow down the root cause by explicitly logging param names that are marked as unused.

Differential Revision: [D29916085](https://our.internmc.facebook.com/intern/diff/D29916085/)

[ghstack-poisoned]
